<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Задача" data-toc-modified-id="Задача-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Изучаем-файл--с-клиентами" data-toc-modified-id="Изучаем-файл--с-клиентами-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучаем файл  с клиентами</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Дерево решений</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Рандомный-лес" data-toc-modified-id="Рандомный-лес-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Рандомный лес</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Логистическая-регресия" data-toc-modified-id="Логистическая-регресия-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Логистическая регресия</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверьте-модели-на-адекватность" data-toc-modified-id="Проверьте-модели-на-адекватность-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Проверьте модели на адекватность</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

## Описание проекта

### Задача

Нужно перевести клиентов с архивных тарифов на новые, для этого нужно посторить модель которая будет предлагать тарифы.
Для тренировки есть база клиентов которые уже перешли с архивных тарифов на "Ултра" и "Смарт". 

### Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- *сalls* — количество звонков,
- *minutes* — суммарная длительность звонков в минутах,
- *messages* — количество sms-сообщений,
- *mb_used* — израсходованный интернет-трафик в Мб,
- *is_ultra* — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Изучаем файл  с клиентами

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier as RFT
from sklearn.tree import DecisionTreeClassifier as DTC
from sklearn.linear_model import LogisticRegression as LR
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

In [2]:
df = pd.read_csv("/datasets/users_behavior.csv")

In [3]:
display(df.info())
df.sample(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


None

calls  minutes  messages   mb_used  is_ultra
2688   57.0   371.55      16.0  12283.46         0
893    55.0   314.72       0.0  17451.24         0
3086   57.0   341.45      73.0  14702.49         1
2481   77.0   551.43      56.0   7694.58         0
851    75.0   529.66      20.0  21749.32         0

### Вывод

Файл извлечен, не имеет пропусков и готов к дальнейшей работе

## Разбейте данные на выборки

Для дальнейшей работы нужно сделать две выборки из базы с клиентами:
1. Тренировочная 75% (df_train)
2. Тестовая 25% (df_test)

In [4]:
df_train, df_test = train_test_split(df, test_size=.25, random_state=46812)

Посмторим что получилось

In [5]:
display(df_train.info())
df_train.sample(5)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2410 entries, 2841 to 1589
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     2410 non-null   float64
 1   minutes   2410 non-null   float64
 2   messages  2410 non-null   float64
 3   mb_used   2410 non-null   float64
 4   is_ultra  2410 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 113.0 KB


None

calls  minutes  messages   mb_used  is_ultra
3045   32.0   191.51      23.0   7154.59         0
421    78.0   483.22      28.0   9210.40         1
1734   73.0   545.96       0.0  12799.60         1
1630   80.0   543.50       3.0  16831.90         0
253   106.0   802.98     171.0  12136.63         1

In [6]:
display(df_test.info())
df_test.sample(5)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 804 entries, 2159 to 486
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     804 non-null    float64
 1   minutes   804 non-null    float64
 2   messages  804 non-null    float64
 3   mb_used   804 non-null    float64
 4   is_ultra  804 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 37.7 KB


None

calls  minutes  messages   mb_used  is_ultra
2838   41.0   276.03      72.0   2621.83         1
1781   36.0   285.92     109.0  19103.31         0
562    35.0   229.53      66.0  17303.54         0
2492   80.0   496.92      19.0  15212.54         0
425    91.0   605.75      36.0   9388.43         0

Теперь нужно из df_train и df_test убрать ответы в отдельную переменную

In [7]:
df_train_features = df_train.drop(["is_ultra"], axis= 1)
df_train_answers = df_train["is_ultra"]
df_test_features = df_test.drop(["is_ultra"], axis= 1)
df_test_answers = df_test["is_ultra"]

### Вывод

Все готово что бы обучать модель

## Исследуйте модели

Поробуем три модели:
1. Дерево решений
2. Случайный лес
3. Логистическая регрессия

### Дерево решений

Для получения наилучшего результата нужно будет перебрать гиперпараметры модели. Для дерева решений это max_depth, min_samples_split и min_samples_leaf.

In [8]:
parametrs = { 'max_depth': range (1,20, 2),
              'min_samples_leaf': range (1,8),
              'min_samples_split': range (2,10,2) }
model_tree = GridSearchCV(DTC(random_state=46812), parametrs)
model_tree.fit(df_train_features, df_train_answers)
model_tree.best_params_

{'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2}

In [21]:
model_tree.best_score_

0.7937759336099586

#### Вывод

Дерево решений при параметрах 'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2 дает правильный ответ на валидационной выборке в 79% случаев

### Рандомный лес

Тут будем перебирать теже гипер параметры, только еще добавится n_estimators

In [10]:
%%time
parametrs = { 'max_depth': range (1,21, 2),
              'min_samples_leaf': range (1,8),
              'min_samples_split': range (2,10,2),
              'n_estimators' : range(1,11)}
model_forest = GridSearchCV(RFT(random_state=46812), parametrs)
model_forest.fit(df_train_features, df_train_answers)
model_forest.best_params_

{'max_depth': 7,
 'min_samples_leaf': 2,
 'min_samples_split': 2,
 'n_estimators': 10}

In [22]:
model_forest.best_score_

0.8087136929460581

#### Вывод

Рандомный лес получил ответ в 81% случаев при параметрах 'max_depth': 7,'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 10

### Логистическая регресия

У данной модели можно поиграться только с одним параметром max_iter.

In [18]:
parametrs = {"solver" : ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'],
             "max_iter": range(1000,2000,100),
             "penalty": ['l1', 'l2', 'elasticnet', 'none']}
model_LR = GridSearchCV(LR(random_state=46812), parametrs)
model_LR.fit(df_train_features, df_train_answers)
model_LR.best_params_

/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 593, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py", line 1306, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py", line 443, in _check_solver
    raise ValueError("Solver %s supports only 'l2' or 'none' penalties, "
ValueError: Solver newton-cg supports only 'l2' or 'none' penalties, got l1 penalty.

  warnings.warn("Estimator fit failed. The score on this train-test"
/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610

{'max_iter': 1000, 'penalty': 'l2', 'solver': 'newton-cg'}

In [23]:
model_LR.best_score_

0.749792531120332

#### Вывод

Логическая регресия получила 75% правильных ответов при 'max_iter': 1000, 'penalty': 'l2', 'solver': 'newton-cg'

## Проверьте модель на тестовой выборке

Проверим все три модели на тестовой выборке и посмотрим на их результаты.

In [24]:
display(accuracy_score(df_test_answers, model_tree.predict(df_test_features)))
display(accuracy_score(df_test_answers, model_forest.predict(df_test_features)))
display(accuracy_score(df_test_answers, model_LR.predict(df_test_features)))

0.7910447761194029

0.8159203980099502

0.7375621890547264

### Вывод

Как и на валидационной выборке рандомный лес показал себя лучше всех тут он достиг 82% правильных ответов, остальные модели показали результат в 79% и 74%

## Проверьте модели на адекватность

По скольку у нас всего 2 варианта ответов то вероятность просто угадать будет 50%. Следоательно раз модель угадывает в 80% случаев то она работает, а не гадает.

## Вывод

Построив три модели с разными гиперпараметрами получили что случайный лес показал лучшие результаты на тестовой выборке 82%. Остальные модели оказались хуже. Так же модель не угадывает результаты, а именно предсказывает, поскольку результаты намного выше случайных.